SO-3933: published vs unpublished snapshot export

cmark opened the pull request:
This PR fixes an RF2 Snapshot export issue when the RF2 export was missing some of the published content when the configuration was not to include pre-release content but the reference branch contained some.

cmark committed:
test(snomed): add test case verifying snapshot export issue

Issue https://snowowl.atlassian.net/browse/SO-3933

cmark committed:
test(snomed): simplify how we report RF2 export errors during tests

cmark committed:
fix(snomed): fall back to the most recent version branch...

...when exporting Snapshot content with `includeUnpublished=false` but
there are unpublished components present on the export reference branch.

apeteri commented on snomed/com.b2international.snowowl.snomed.datastore/src/com/b2international/snowowl/snomed/datastore/request/rf2/SnomedRf2ExportRequest.java:
```diff
@@ -484,7 +486,15 @@ private List<String> computeBranchesToExport(final String referenceBranch, final
+					if (versionsToExport.isEmpty()) {
+						throw new BadRequestException("Snapshot export without unpublished components requires at least one version.");
+					}
+					branchesToExport.add(versionsToExport.first().getBranchPath());
```
`versionsToExport` is a `TreeSet` with an ascending effective time sort order. I think we need the element with the greatest value here, not the smallest.

To confirm, add at least one more version to the test case that makes a different change on the component.

cmark replied:
Yes, you are right I thought that is a TreeSet but the versions are in reverse order because of the ordering we use when fetching the versions. The Ordering at the top of the class, unfortunately, orders them by chronological effective time order. To reflect this, I fixed the issue, renamed the variable, and updated the test case with an additional version and component.

cmark committed:
refactor(commons): eliminate unnecessary Pair implementations

Support only a single Pair value type class with serialization and
equals/hashCode support.

cmark committed:
fix(snomed): invalid selection of first version instead of the latest...

...when exporting Snapshot without unpublished content

Rename `versionsToExport` variable to
`versionsToExportInChronologicalOrder` to reflect our intention better.

apeteri reviewed (approved):
LGTM!

nagyo reviewed (approved):
🍏
